### PR TITLE
Fix #173: Support multiple cron expressions per DAG

### LINKED ISSUE #173: Support multiple schedules
* [ ] Make `SCHEDULE` accept either a single cron expression or an array e.g.:
```
SCHEDULE=[
  "0 7,17,19,21 * * MON-FRI", 
  "0 17 * * MON-FRI"
]
```

* [ ] Update UI to handle above

* [ ] Go through our DAGs (`.ini`) files and cleanup all dupe jobs

### PR DESCRIPTION
**Pull request by vdat-coatue**
- allows use of multi schedules via stringified json: ex: `'["* * * * *", "*/5 * * * * *"]'`
- has backward compatibility with current string schedule syntax, ex: `"*/5 * * * * "`
- Tested Kinds of schedules: None, single, and multiple
- Test cron expressions: 
    - @ expressions like `@hourly`
    - relative `*/5 * * * *`
    - absolute `0 8 * * MON-FRI`

Updates UI can be viewed here: https://github.com/coatue/datasyncs/pull/1330

### CODE REVIEW

File: airflow/models/__init__.py [hunk 1]
```diff
@@ -3299,60 +3311,144 @@ def following_schedule(self, dttm, tz=None):
...
+    #         else:
+    #             # absolute (e.g. 3 AM)
+    #             naive = cron.get_next(datetime)
+    #             self.log.info(f"In following_schedule naive={naive}")
+    #             tz = pendulum.timezone(self.timezone.name)
+    #             following = timezone.make_aware(naive, tz)
+    #             self.log.info(f"In following_schedule following={following}")
+    #         return timezone.convert_to_utc(following)
+    #     elif self._schedule_interval is not None:
+    #         return dttm + self._schedule_interval
```
Comment (vdat-coatue) [inline airflow/models/__init__.py]: will clean up commentd out code

File: airflow/models/__init__.py [hunk 2]
```diff
@@ -3399,7 +3442,8 @@ def get_next_run_date(self, after_date):
             return None
         if isinstance(self._schedule_interval, six.string_types):
             # six.string_types is for compatibility with python 2 and 3, XXX do we need compatibility?
-            return self.following_schedule(after_date)
+            # following_schedules will return a list now, need only the head element
+            return self.following_schedule(after_date)[0]
```
Comment (ikurian-coatue) [inline airflow/models/__init__.py]: do we need to sort this before taking head ?

Comment (vdat-coatue) [reply airflow/models/__init__.py]: good point, i'm sorting other places. Will update here

Comment (vdat-coatue) [reply airflow/models/__init__.py]: nvm, this only will return a single element, so need need to sort.

File: airflow/models/__init__.py [hunk 3]
```diff
@@ -3299,60 +3310,78 @@ def following_schedule(self, dttm, tz=None):
...
-                # absolute (e.g. 3 AM)
-                naive = cron.get_prev(datetime)
-                tz = pendulum.timezone(self.timezone.name)
-                previous = timezone.make_aware(naive, tz)
-            return timezone.convert_to_utc(previous)
-        elif self._schedule_interval is not None:
-            return dttm - self._schedule_interval
+
+        previous_schedules = []
+        intervals = []
```
Comment (pathikrit) [inline airflow/models/__init__.py]: DRY: Lines 3313-3342 and lines 3353-3384 are pretty much same

File: airflow/models/__init__.py [hunk 4]
```diff
@@ -3419,9 +3449,14 @@ def get_next_run_date(self, after_date):
             next_date = after_date + schedule_interval - (after_date - start_date) % schedule_interval
             """
             start_date = min(t.start_date for t in self.tasks)
-            return after_date \
-                   - timedelta(seconds=((after_date - start_date).total_seconds() % self._schedule_interval.total_seconds())) \
-                   + self._schedule_interval
+            try:
+                intervals = json.loads(self._schedule_interval)
+                following_schedules = self.following_schedule(after_date)
+                return sorted(following_schedules)[0]
```
Comment (pathikrit) [inline airflow/models/__init__.py]: `min(foo)` is better than `sorted(foo)[0]`:

```suggestion
                return min(following_schedules)
```

File: airflow/www/templates/airflow/coatue-dags.html [hunk 5]
```diff
@@ -166,10 +166,12 @@ <h2>DAGs</h2>
 
     <!-- Column 7: Dag Schedule -->
     <td>
-      <a class="label label-default schedule {{ dag.dag_id }}" href="{{ url_for('dagrun.index_view') }}?flt2_dag_id_equals={{ dag.dag_id }}"
-         data-toggle="tooltip" title = "{{ get_human_readable_cron(dag.schedule_interval, dag_obj.timezone.name) | string }}">
-        {{ dag.schedule_interval }}
-      </a>
+      {% for schedule_interval in dag.intervals %}
```
Comment (pathikrit) [inline airflow/www/templates/airflow/coatue-dags.html]: Can you post a screenshot in the PR body like we do for other UX changes?

File: airflow/jobs.py [hunk 6]
```diff
@@ -853,30 +853,37 @@ def create_cron_dag_run(self, dag, session=None):
...
+            self.log.info(
+                f"next_run_date type={type(nrd)} end type={type(end)} next_run_date < end: {nrd < end}")
+            if nrd < end:
+                last_scheduled_run = self.get_last_scheduled_run(dag, session)
+                if not last_scheduled_run or last_scheduled_run < nrd:
+                    #  create run
+                    next_dag_run = dag.create_dagrun(
+                        run_id=DagRun.ID_PREFIX + nrd.isoformat(),
+                        execution_date=nrd,
+                        start_date=timezone.utcnow(),
```
[the remark this answers is not in the document]
Comment (vdat-coatue) [reply airflow/jobs.py]: unfortunately its not available until we merge the repos. But I will add a todo

File: airflow/jobs.py [hunk 7]
```diff
@@ -853,30 +853,37 @@ def create_cron_dag_run(self, dag, session=None):
...
             else:
-                self.log.info(f"CRON scheduler: Found last scheduled run: {last_scheduled_run} greater than or equal to next run date {next_run_date}, SKIPPING")
-                return
+                self.log.info(f"CRON scheduler: Next run is at {nrd}, outside Recovery window, SKIPPING")
+
+        if candidates:
+            return sorted(candidates)[0]
         else:
-            self.log.info(f"CRON scheduler: Next run is at {next_run_date}, outside Recovery window, SKIPPING")
             return
```
[the remark this answers is not in the document]
Comment (vdat-coatue) [reply airflow/jobs.py]: doesn't work with return
``` python
>>> def test(h):
...     return min(h) if h else
  File "<stdin>", line 2
    return min(h) if h else
```

File: airflow/www/templates/airflow/coatue-dags.html (hunk 5 above)
Comment (vdat-coatue) [reply airflow/www/templates/airflow/coatue-dags.html]: https://github.com/coatue/datasyncs/pull/1330

File: airflow/settings.py [hunk 8]
```diff
@@ -42,6 +42,8 @@
 RBAC = conf.getboolean('webserver', 'rbac')
 
 TIMEZONE = pendulum.timezone('UTC')
```
Comment (pathikrit) [inline airflow/settings.py]: out of scope but a todo to rename line 44 to `UTC_TZ` would be nice

File: airflow/jobs.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (pathikrit) [reply airflow/jobs.py]: ok:
```
   return min(candidates) if candidates else None
```

File: airflow/models/__init__.py [hunk 9]
```diff
@@ -3292,67 +3306,81 @@ def is_fixed_time_schedule(self):
 
         return False
 
+
+    def calculate_schedule(self, dttm, delta_lambda, schedule_lambda, cron_lambda, default_lambda, tz=None):
```
Comment (skim-coatue) [inline airflow/models/__init__.py]: Can you clarify what the parameters of this function are in the docstring?

File: airflow/utils/cron.py [hunk 10]
```diff
@@ -68,4 +68,11 @@ def __str__(self):
...
 
 def get_human_readable_cron(expression: str, timezone: str, include_timezone: bool = True) -> str:
-    return str(CronExpression(expression = expression, timezone = timezone, show_timezone = timezone != 'US/Eastern'))
+    intervals = []
+    try:
+        intervals = json.loads(expression)
+    except Exception as e:
+        intervals = [expression]
+
+    return " and ".join( [str(CronExpression(expression = expression, timezone = timezone, show_timezone = timezone != 'US/Eastern')) for expression in intervals] )
```
Comment (skim-coatue) [inline airflow/utils/cron.py]: Nit: Does this function need to be refactored? It looks like we're passing in each cron separately in `coatue-dags.html`

Comment (vdat-coatue) [reply airflow/utils/cron.py]: undid refactor

**Pull request merged**